modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessSuggest.java

```diff
@@ -0,0 +1,251 @@
+import java.util.function.Function;
+
+/**
+ * PainlessSuggest generates suggestions for partially completed scripts
+ * based on context and source. The parsing for suggestions is
+ * extremely lenient and never issues an error outright. The parser attempts
+ * to recover given extraneous or unknown input. If an error is unrecoverable,
+ * no suggestions are given. Also note that no suggestions are given if
+ * a type is resolved as def as there is no way to know what def represents.
+ */
+public class PainlessSuggest {
+
+    /**
+     * TokenState keeps track of the current token. The state machines
+     * each increment a single token at a time and make decisions by
+     * storing state appropriately.
+     */
+    protected static class TokenState {
+
+        protected final List<? extends Token> tokens;
+
+        protected int current = 0;
+
+        protected TokenState(List<? extends Token> tokens) {
+            this.tokens = Collections.unmodifiableList(tokens);
+        }
+    }
+
+    /**
+     * FunctionMachine collects information about user-defined functions.
+     * This is required for suggestions relating to static-style calls,
+     * and to determine what top-level variables are available.
+     */
+    protected static class FunctionMachine {
+
+        protected static class FunctionData {
+
+            protected String returnType = "";
+            protected String functionName = "";
+            protected final List<String> parameterTypes = new ArrayList<>();
+            protected final List<String> parameterNames = new ArrayList<>();
+
+            protected int bodyStartToken = -1;
+            protected int bodyEndToken = -1;
+
+            @Override
+            public String toString() {
+                return "FunctionState{" +
+                        "returnType='" + returnType + '\'' +
+                        ", functionName='" + functionName + '\'' +
+                        ", parameterTypes=" + parameterTypes +
+                        ", parameterNames=" + parameterNames +
+                        ", bodyStartToken=" + bodyStartToken +
+                        ", bodyEndToken=" + bodyEndToken +
+                        '}';
+            }
+        }
+
+        protected static class FunctionState extends TokenState {
+
+            protected int target = 0;
+
+            protected String returnType;
+            protected String functionName;
+            protected String parameterType;
+            protected FunctionData functionData;
+
+            protected final List<FunctionData> functions = new ArrayList<>();
+
+            protected int brackets;
+
+            protected FunctionState(List<? extends Token> tokens) {
+                super(tokens);
+            }
+        }
+
+        protected static final List<Function<FunctionState, Integer>> fstates;
+
+        static {
+            fstates = new ArrayList<>();
+            // 0 - possible start of function
+            fstates.add(fs -> {
```

Comment:
Can we have an Enum for the states and a function per Enum that takes a token and returns the next state?